docs/spec.md

```diff
@@ -1136,34 +1149,40 @@ def batch_norm_inference(operand, scale, offset, mean, variance, epsilon, featur
   return add(multiply(scale_bcast, normalized_operand), offset_bcast)
 ```
 
+For quantized types, performs
+`dequantize_op_quantize(lambda operand, scale, offset, mean, variance:
+batch_norm_inference(operand, scale, offset, mean, variance, epsilon,
+feature_index), operand, scale, offset, mean, variance, type(result))`.
+
 #### Inputs
 
-| Label | Name            | Type                                        | Constraints   |
-|-------|-----------------|---------------------------------------------|---------------|
-| (I1)  | `operand`       | tensor of floating-point type               | (C1-C7)       |
-| (I2)  | `scale`         | 1-dimensional tensor of floating-point type | (C2), (C3)    |
-| (I3)  | `offset`        | 1-dimensional tensor of floating-point type | (C2), (C4)    |
-| (I4)  | `mean`          | 1-dimensional tensor of floating-point type | (C5)          |
-| (I5)  | `variance`      | 1-dimensional tensor of floating-point type | (C2), (C6)    |
-| (I6)  | `epsilon`       | constant of type `f32`                      |               |
-| (I7)  | `feature_index` | constant of type `si64`                     | (C1), (C3-C6) |
+| Label | Name            | Type                                                                | Constraints   |
+|-------|-----------------|---------------------------------------------------------------------|---------------|
+| (I1)  | `operand`       | tensor of floating-point type or per-tensor quantized tensor        | (C1-C7)       |
+| (I2)  | `scale`         | 1-dimensional tensor of floating-point or per-tensor quantized type | (C2), (C3)    |
+| (I3)  | `offset`        | 1-dimensional tensor of floating-point or per-tensor quantized type | (C2), (C4)    |
+| (I4)  | `mean`          | 1-dimensional tensor of floating-point or per-tensor quantized type | (C5)          |
+| (I5)  | `variance`      | 1-dimensional tensor of floating-point or per-tensor quantized type | (C2), (C6)    |
+| (I6)  | `epsilon`       | constant of type `f32`                                              |               |
```

Comment:
That is a good point!
The reason epsilon is not provided in  quantized form is because it is a program constant and it would be non trivial to determine the quantization parameters (scale & zero point) for its quantized form.  For example, during the compilation time we do not know  the quantization parameters of the variance. So, the the quantization parameter to be selected for the epsilon would be arbitrary and the op semantics should involve re-scaling it to the quantization parameter of the variance. 

In any case, let me revisit point as part of https://github.com/openxla/stablehlo/issues/1404., I have added an item in that ticket.

Reply:
While I think it is fine to allow quantized type for `epsilon` (in other words, no objection from my side), it will be difficult to use.  The purpose of `epsilon` is to avoid division by zero when we calculate `1 / sqrt(variance + epsilon)`.  The model author must knowingly choose a constant value that won't be rounded to zero when `epsilon` is quantized.  Furthermore, the difference between the `scale`/`zero_point` of `variance` and `epsilon` should be small enough to prevent `epsilon` from being ignored.  And, the quantizer must ensure the quantized type of the `sqrt(variance + epsilon)` has enough resolution and won't round down the result value to zero.

In general, I question the feasibility of quantizing normalization-like operations by replacing the types of the floating point normalization implementations.  But that topic will be far beyond of the scope of this CL.  I am fine with either option: (1) submit as-is, (2) allow all operands to get quantized type (and leave the problem to the quantizer), or (3) scope out normalization-like op for further discussion.  All of these options are fine with me.

Reply:
Thanks for the comments. I think this is fine to be tracked as part of #1404 as it is integer only computation that will be affected by this. So fine with me to submit this and take further discussion in #1404.